File: 二分查找/162.寻找峰值.py

class Solution1:
    def findPeakElement(self, nums: List[int]) -> int:
        l, r = 0, len(nums) - 1

        # 峰值在比nums[m]大的一侧
        while l <= r:
            m = l + (r - l) // 2
            if m > 0 and nums[m] < nums[m - 1]: # m = 0 时会越界
                r = m - 1
            elif m < len(nums) - 1 and nums[m] < nums[m + 1]: # m = len(nums) - 1 时会越界
                l = m + 1
            else:
                return m

# 二分查找 少判断很多边界情况
class Solution2:
    def findPeakElement(self, nums: List[int]) -> int:
        l, r = 0, len(nums) - 1
        while l + 1 < r:
            m = l + (r - l) // 2
            if nums[m] < nums[m + 1]:
                l = m
            else:
                r = m

        if nums[l] > nums[r]:
            return l
        return r

File: 二分查找/162.寻找峰值.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findPeakElement = function (nums) {
  let l = 0, r = nums.length - 1;

  while (l < r) { //! 不是<= 如果是<=就会越界
    const m = (l + r) >> 1;  //* 正整数右移向下取整
    if (nums[m] < nums[m + 1]) {
      l = m + 1;
    } else {
      r = m;  // 如果前面在减小, 峰值可能是m也可能是m-1
    }
  }
  return l;
};
